### PR TITLE
[DMS-71] Selectively allow natural key updates

### PR DESCRIPTION
**Pull request by simpat-adam**

### CODE REVIEW

File: src/tests/EdFi.DataManagementService.Tests.E2E/Features/Resources/UpdateResourcesValidation.feature [hunk 1]
```diff
@@ -534,3 +534,34 @@ Feature: Resources "Update" Operation validations
                         "correlationId": null
                   }
                   """
+
+        Scenario: 18 Verify clients can update the identity of resources that allow identity updates
```
Comment (stephenfuqua) [inline src/tests/EdFi.DataManagementService.Tests.E2E/Features/Resources/UpdateResourcesValidation.feature:538]: 👍

**Review by bradbanister — CHANGES_REQUESTED** [review]
`AllowIdentityUpdateOverrides` is not a database-level override. It should be overriding the `ResourceInfo.AllowIdentityUpdates` boolean field that is part of, for example, `IUpdateRequest`. Suggest doing this in `BuildResourceInfoMiddleware`.

File: src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/UpdateDocumentById.cs [hunk 2]
```diff
@@ -93,14 +95,43 @@ NpgsqlTransaction transaction
...
-                );
-                return new UpdateResult.UpdateFailureImmutableIdentity(
-                    $"Identifying values for the {updateRequest.ResourceInfo.ResourceName.Value} resource cannot be changed. Delete and recreate the resource item instead."
-                );
+                // Extracted referential id does not match stored. Must be attempting to change identity.
+                if (updateRequest.ResourceInfo.AllowIdentityUpdates || allowIdentityUpdateOverrides.Contains(updateRequest.ResourceInfo.ResourceName.Value))
+                {
+                    // Identity update is allowed
+                    _logger.LogInformation(
+                        "Updating Natural Key - {TraceId}",
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/UpdateDocumentById.cs:103]: Use "Updating identity" not "Updating Natural Key"

File: src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/UpdateDocumentById.cs [hunk 3]
```diff
@@ -93,14 +95,43 @@ NpgsqlTransaction transaction
...
+                            updateRequest.TraceId
+                        );
+                        return new UpdateResult.UpdateFailureNotExists();
+                    }
+                }
+                else
+                {
+                    // Identity update not allowed
+                    _logger.LogInformation(
+                        "Failure: Natural key does not match on update - {TraceId}",
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/UpdateDocumentById.cs:128]: Change to "identity" vs "natural key"

File: src/backend/EdFi.DataManagementService.Backend.Postgresql.Test.Integration/UpdateTests.cs [hunk 4]
```diff
@@ -157,6 +157,59 @@ public async Task It_should_not_have_changed_the_document()
...
+                .GetById(
+                    CreateGetRequest(_defaultResourceName, _documentUuidGuid),
+                    Connection!,
+                    Transaction!
+                );
+            (getResult as GetResult.GetSuccess)!.EdfiDoc.ToJsonString().Should().Contain("\"abc\":2");
+        }
+    }
+
+
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql.Test.Integration/UpdateTests.cs:212]: Need a test where something references the document before the identity change.

File: src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/SqlAction.cs [hunk 5]
```diff
@@ -467,6 +467,56 @@ NpgsqlTransaction transaction
...
+        NpgsqlConnection connection,
+        NpgsqlTransaction transaction
+    )
+    {
+        var result = await GetPostgresExceptionRetryPipeline().ExecuteAsync(async _ =>
+        {
+            try
+            {
+                await using var command = new NpgsqlCommand(
+                    @"UPDATE dms.Alias AS a
```
Comment (bradbanister) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/SqlAction.cs:488]: Does this work if a document references the document before the identity change? Doesn't the `Reference` table need updating? I don't see an update cascade setting on the FK in the 1000 SQL script, maybe that's all you need?

**Review by bradbanister — APPROVED** [review]

**Review by bradbanister — APPROVED** [review]

**Pull request merged**